ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -0,0 +1,489 @@
+    location->function.name.ptr = frame->name;
+    location->function.name.len = strlen(frame->name);
+    location->function.filename.ptr = frame->filename;
+    location->function.filename.len = strlen(frame->filename);
+    location->line = frame->line;
+  }
+
+  stack_iteration_data stack_data;
+  stack_data.inuse_objects = record->inuse_objects;
+  stack_data.locations = (ddog_prof_Slice_Location) {.ptr = locations, .len = stack->frames_len};
+
+  iteration_data->for_each_callback(stack_data, iteration_data->for_each_callback_extra_arg);
+
+  return ST_CONTINUE;
+}
+
+void heap_recorder_iterate_stacks(heap_recorder *heap_recorder, void (*for_each_callback)(stack_iteration_data stack_data, void *extra_arg), void *for_each_callback_extra_arg) {
+  pthread_mutex_lock(&heap_recorder->records_mutex);
+  internal_iteration_data internal_iteration_data;
+  internal_iteration_data.for_each_callback = for_each_callback;
+  internal_iteration_data.for_each_callback_extra_arg = for_each_callback_extra_arg;
+  internal_iteration_data.heap_recorder = heap_recorder;
+  st_foreach(heap_recorder->heap_records, st_heap_records_iterate, (st_data_t) &internal_iteration_data);
+  pthread_mutex_unlock(&heap_recorder->records_mutex);
+}
+
+void commit_allocation(heap_recorder *heap_recorder, heap_stack *heap_stack, VALUE obj, unsigned int weight) {
+  heap_record *heap_record = NULL;
+  if (!st_lookup(heap_recorder->heap_records, (st_data_t) heap_stack, (st_data_t*) &heap_record)) {
+    heap_record = heap_record_init(heap_stack);
+    if (st_insert(heap_recorder->heap_records, (st_data_t) heap_stack, (st_data_t) heap_record)) {
+      rb_raise(rb_eRuntimeError, "Duplicate heap stack tracking: %p", heap_stack);
+      return;
+    };
+  } else {
+    // FIXME: Figure out a way to not have to instantiate a new stack only to free it if it's
+    // already sampled. Something like supporting indexing the heap_records table with
+    // ddog_prof_Slice_Location objects directly for instance.
+    heap_stack_free(heap_stack);
+  }
+
+  object_record *object_record = object_record_init(obj, weight, heap_record);
+  if (st_insert(heap_recorder->object_records, (st_data_t) obj, (st_data_t) object_record) != 0) {
+    // Object already tracked?
+    object_record_free(object_record);
+    rb_raise(rb_eRuntimeError, "Duplicate heap object tracking: %lu", obj);
+    return;
+  }
+
+  fprintf(stderr, "Committed allocation of %lu (heap_record=%p, object_record=%p)\n", obj, heap_record, object_record);
+
+  heap_record->num_tracked_objects++;
+  heap_record->inuse_objects += weight;
+}
+
+void commit_free(heap_recorder *heap_recorder, VALUE obj) {
+  st_data_t key = (st_data_t) obj;
+  object_record *object_record = NULL;
+  if (!st_delete(heap_recorder->object_records, (st_data_t*) &key, (st_data_t*) &object_record)) {
+    // Object not tracked?
+    rb_raise(rb_eRuntimeError, "Committing free of untracked object");
+    return;
+  }
```

Comment:
This is even weirder, since it gets called from `record_heap_free` and that one already looks up in the map and checks that there's something there, so it would be really weird to see it fail here.

Note that if it turns out to be the case that the `RUBY_INTERNAL_EVENT_FREEOBJ` tracepoint gets called during GC, this one will make the VM explode since we're not allowed to allocate during GC (and we're allocating an exception).

P.s.: No need for `return` after raise, since it'll longjmp out :)